pkg/urbit/noun/events.c

```diff
@@ -937,7 +900,96 @@ u3e_save(void)
   _ce_patch_free(pat_u);
   _ce_patch_delete();
 
-  _ce_backup();
+  c3_c pax_c[8193];
+  snprintf(pax_c, sizeof(pax_c), "%s/.urb/bhk", u3P.dir_c);
+  if ( c3n == u3e_copy(pax_c) ) {
+    fprintf(stderr, "loom: failed to copy snapshot to %s\r\n", pax_c);
+  }
+}
+
+//! @n (1) Attempt to `dir_c`.
+//! @n (2) Attempt to create north image file in `dir_c`.
+//! @n (3) Attempt to create south image file in `dir_c`.
+//! @n (4) Copy north and south image files to `dir_c` from `u3P.dir_c`.
+c3_o
+u3e_copy(const c3_c* const dir_c)
+{
+  c3_o suc_o = c3n;
+  static c3_i fla_i = O_RDWR | O_CREAT;
+  static const mode_t mod_u = 0666;
+
+  // (1)
+  if ( 0 != mkdir(dir_c, 0700) && EEXIST != errno ) {
+    fprintf(stderr, "loom: failed to create %s: %s\r\n",
+            dir_c, strerror(errno));
+    goto exit;
+  }
+
+  // (2)
+  u3e_image nop_u = { .nam_c = nor_nam_c, .pgs_w = 0 };
+  c3_c pan_c[8193];
+  snprintf(pan_c, sizeof(pan_c), "%s/%s", dir_c, nop_u.nam_c);
+  if ( -1 == (nop_u.fid_i = open(pan_c, fla_i, mod_u)) ) {
+    fprintf(stderr, "loom: failed to open %s: %s\r\n", pan_c, strerror(errno));
+    goto exit;
+  }
+
+  // (3)
+  u3e_image sop_u = { .nam_c = sou_nam_c, .pgs_w = 0 };
+  c3_c pas_c[8193];
+  snprintf(pas_c, sizeof(pas_c), "%s/%s", dir_c, sop_u.nam_c);
+  if ( -1 == (sop_u.fid_i = open(pas_c, fla_i, mod_u)) ) {
+    fprintf(stderr, "loom: failed to open %s: %s\r\n", pas_c, strerror(errno));
+    goto close_north;
+  }
+
+  // (4)
+  if ( (c3y == _ce_image_copy(&u3P.nor_u, &nop_u)) &&
+       (c3y == _ce_image_copy(&u3P.sou_u, &sop_u)) )
+  {
+    suc_o = c3y;
+  }
+
+close_south:
+  close(sop_u.fid_i);
+  if ( c3n == suc_o ) {
+    unlink(pas_c);
+  }
+close_north:
+  close(nop_u.fid_i);
+  if ( c3n == suc_o ) {
+    unlink(pan_c);
+  }
+exit:
+  return suc_o;
+}
+
+//! @n (1)
+c3_o
+u3e_load(const c3_c* const dir_c)
+{
+  if ( !dir_c ) {
+    return c3n;
+  }
+
+  u3e_image nor_u, sou_u;
+
+  if ( (c3n == _ce_image_open(dir_c, &nor_u)) ||
+       (c3n == _ce_image_open(dir_c, &sou_u)) )
+  {
+    fprintf(stderr, "boot: failed to load snapshot in %s\r\n", dir_c);
+    exit(1);
+  }
```

Comment:
The error handling/return behavior should be reviewed (or revisited later, it's not the only such issue in this file).